### ISSUE TITLE
Shape inference fails if an unused node is added

### ISSUE THREAD

**Issue by xadupre** [microsoft/onnxruntime#11614]
**Describe the bug**

Two models are alsmot equal. The second one contains two unused nodes. However, onnxruntime can load the first one and execute it. It cannot load the second due to:

```
FAIL : Node () Op (If) [TypeInferenceError] Graph attribute inferencing failed: Node () Op (Slice) [ShapeInferenceError] Input axes has invalid data
```

The error happens when onnxruntime is calling onnx to run shape inference.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows
- ONNX Runtime installed from (source or binary): source
- ONNX Runtime version: github
- Python version: 3.9
- Visual Studio version (if applicable): 2019

**To Reproduce**

Run InferenceSession with the following inputs and the attached files.

```python
inputs = {'x': numpy.random.randn(3, 4, 5, 1).astype(numpy.float32),
          'fft_length': numpy.array([5], dtype=numpy.int64),
          'onesided': numpy.array([0], dtype=numpy.int64),
          'inverse': numpy.array([0], dtype=numpy.int64),
          'normalize': numpy.array([0], dtype=numpy.int64)}
```

Models:

[data.zip](https://github.com/microsoft/onnxruntime/files/8765208/data.zip)

**Expected behavior**

onnxruntime should return the same results with both models.

**Comment by hariharans29** [comment]
Can you elaborate on why the nodes are "unused" in the second model ? You mean it is in a conditional branch of the `If` node that will never be executed (because the condition to the `If` will never be equal what it should be to execute the "unused" nodes ?)

Even if they will be "unused" at runtime for the afore-mentioned reason, ORT still needs to load the graph and it seems to have triggered a sanity check failure in the ONNX shape inference while loading the graph (https://github.com/onnx/onnx/blob/a525f98d5ed31660b629bab90c680a39658a5c08/onnx/defs/tensor/defs.cc#L931) that is complaining that one/more axis values is out of range for the deduced input rank. This *usually* points to something faulty in the model.

TL;DR: Even if a sub-graph will never be used at runtime, it still needs to be "valid" for ORT to be able to load it.

**Comment by hariharans29** [comment]
If you think this is a problem with shape inference, then this issue should be opened in the ONNX repo.

**Comment by gramalingam** [comment]
I wonder if the unused nodes are misleading. Maybe the scalar shape specified for the inputs is the main cause (regardless of what the bug is, if any). It might help to test with the input shape specified, but include the unused nodes. If it still fails, at least we know which of the two deltas is causing the problem.

**Comment by gramalingam** [comment]
Ok, the problem is with the input shapes specified. You have used an empty shape for the input and output, which makes it a scalar. And you cannot slice a scalar. You should change the input shapes to have a symbolic dimension, some shape like [N] or [M, N] depending on the rank of the input/output.

**Comment by xadupre** [comment]
The rank is unknown is that case.
